AppDB/backup/cassandra_backup.py

```diff
@@ -0,0 +1,308 @@
+import os
+import sys
+import tarfile
+import subprocess
+from subprocess import call
+from subprocess import CalledProcessError
+
+import backup_exceptions
+import gcs_helper
+from backup_recovery_constants import StorageTypes
+
+sys.path.append(os.path.join(os.path.dirname(__file__), "../"))
+import dbconstants
+
+sys.path.append(os.path.join(os.path.dirname(__file__), "../../lib/"))
+import constants
+
+sys.path.append(os.path.join(os.path.dirname(__file__), "../cassandra/"))
+import start_cassandra
+import shut_down_cassandra
+from cassandra_interface import NODE_TOOL
+from cassandra_interface import KEYSPACE
+
+# Location where we place the tar of the snapshot.
+BACKUP_DIR_LOCATION = "/opt/appscale/backups"
+
+# File location of where the latest backup goes.
+BACKUP_FILE_LOCATION = "{0}/cassandra_backup.tar.gz".format(BACKUP_DIR_LOCATION)
+
+# Cassandra directories to remove to get rid of data.
+CASSANDRA_DATA_SUBDIRS = ["Keyspace1", "system",
+  # TODO Is the rest needed?
+  # "commitlog", "saved_caches",
+  # "system_traces"
+]
+
+def clear_old_snapshots():
+  """ Remove any old snapshots to minimize diskspace usage both locally. """
+  logging.info('Removing old Cassandra snapshots...')
+  try:
+    call([NODE_TOOL, 'clearsnapshot'])
+  except CalledProcessError as error:
+    logging.error('Error while deleting old Cassandra snapshots. Error: {0}'.\
+      format(str(error)))
+
+def create_snapshot(snapshot_name=''):
+  """ Perform local Cassandra backup by taking a new snapshot.
+
+  Args:
+    snapshot_name: A str, optional. A fixed name for the snapshot to create.
+  Returns:
+    True on success, False otherwise.
+  """
+  logging.info('Creating new Cassandra snapshots...')
+  try:
+    call([NODE_TOOL, 'snapshot'])
+  except CalledProcessError as error:
+    logging.error('Error while creating new Cassandra snapshots. Error: {0}'.\
+      format(str(error)))
+    return False
+  return True
+
+def get_cassandra_snapshot_file_names():
+  """ Yields all file names which should be tar'ed up.
+
+  Returns:
+    A list of files.
+  """
+  file_list = []
+  data_dir = "{0}{1}".format(constants.APPSCALE_DATA_DIR, "cassandra")
+  for full_path, _, _ in os.walk(data_dir):
+    if 'snapshots' in full_path:
+      file_list.append(full_path)
+
+  logging.debug("List of snapshot paths: {0}".format(file_list))
+  return file_list
+
+def tar_backup_files(file_paths):
+  """ Tars all snapshot files for a given snapshot name.
+
+  Args:
+    file_paths: A list of files to tar up.
+  Returns:
+    The path to the tar file, None otherwise.
+  """
+
+  # Create backups dir if not there.
+  try:
+    call(["mkdir", "-p", BACKUP_DIR_LOCATION])
+  except CalledProcessError as error:
+    logging.error("Error while creating dir '{0}'. Error: {1}".
+      format(BACKUP_DIR_LOCATION, str(error)))
+    return None
+
+  backup_file_location = BACKUP_FILE_LOCATION
+
+  # Delete previous backup.
+  try:
+    call(["rm", "-f", backup_file_location])
```

Comment:
It's deleted only if the storage backend is remote (like GCS).
For local backups, we don't delete it.